### ISSUE TITLE
Expose environment ids?

### ISSUE THREAD

**Issue by annevk** [slightlyoff/ServiceWorker#643]
Now that we have IDs to identify environments, perhaps environments should have access to them too? E.g. might be useful for sites that allow different session per open document.

Also, we should probably say that environment ids should be unique per origin, but should not be globally unique or something like that, so they can not really be used as additional entropy to identify the user.

**Comment by jakearchibald** [comment]
Is this satisfied by `client.id`?

**Comment by annevk** [comment]
That doesn't give a client access to its own ID, does it?

**Comment by jakearchibald** [comment]
I think we're going to try moving the clients API into all client types. Would `clients.thisId` or something similar solve this?

**Comment by annevk** [comment]
Each global has .clients I guess? Seems a little weird to put it there.

**Comment by wanderview** [comment]
It would be nice we could provide an API to snapshot the current window as a full `Client` snapshot object.

What about something like:

```
let selfClient = new Client(window);

let frame = document.createElement(iframe);
document.body.appendChild(frame);

let FrameClient = new Client(frame.contentWindow);
```

Seems if you have synchronous access to a window global we could simply mint the full Client object immediately.

**Comment by annevk** [comment]
No need to pass in the global then. You could just use the constructor from the correct global. Anyway, that seems reasonable.

**Comment by annevk** [comment]
Although, shouldn't it be asynchronous? Or is synchronous for the current global okay?

**Comment by wanderview** [comment]
> No need to pass in the global then. You could just use the constructor from the correct global. 

I thought it might be nice to be able to specify a same-origin iframe global, etc.  But sure.  We could also have a snapshot method or something that returns a Client.

> Although, shouldn't it be asynchronous? Or is synchronous for the current global okay?

I think synchronous from the current global should be ok.  AFAIK all the values on client are available synchronously in other places.  For example, `document.visibilityState` and `document.hasFocus()`.

**Comment by annevk** [comment]
Yes, for the iframe case you can just do `new self[0].Client()`.

**Comment by monica-ch** [comment]
Bug triage: Need to discuss in TPAC 2025

**Comment by monica-ch** [comment]
Discussed in [TPAC 2025](https://docs.google.com/document/d/1fOjXiGUFf_krkfYaWf5drbqzY7CTxn7EcDYr5ZWjzFg/edit?tab=t.0#heading=h.d1sstrfnte0b) and reached consensus on exposing only Id is safer and simpler.

Broader Clients API exposure can be revisited later if strong use cases emerge.
